pkg/proxy/runner/bounded_frequency_runner.go

```diff
@@ -21,188 +21,141 @@ import (
-// - fn will have no more than 10 seconds between runs
-//
-// The maxInterval must be greater than or equal to the minInterval,  If the
-// caller passes a maxInterval less than minInterval, this function will panic.
-func NewBoundedFrequencyRunner(name string, fn func(), minInterval, maxInterval time.Duration, burstRuns int) *BoundedFrequencyRunner {
-	timer := &realTimer{timer: time.NewTimer(0)} // will tick immediately
-	<-timer.C()                                  // consume the first tick
-	return construct(name, fn, minInterval, maxInterval, burstRuns, timer)
+// `maxInterval` must be greater than or equal to `minInterval`; otherwise,
+// this function will panic.
+func NewBoundedFrequencyRunner(name string, fn func(), minInterval, retryInterval, maxInterval time.Duration) *BoundedFrequencyRunner {
+	return construct(name, fn, minInterval, retryInterval, maxInterval, clock.RealClock{})
 }
 
 // Make an instance with dependencies injected.
-func construct(name string, fn func(), minInterval, maxInterval time.Duration, burstRuns int, timer timer) *BoundedFrequencyRunner {
+func construct(name string, fn func(), minInterval, retryInterval, maxInterval time.Duration, clock clock.Clock) *BoundedFrequencyRunner {
 	if maxInterval < minInterval {
 		panic(fmt.Sprintf("%s: maxInterval (%v) must be >= minInterval (%v)", name, maxInterval, minInterval))
 	}
-	if timer == nil {
-		panic(fmt.Sprintf("%s: timer must be non-nil", name))
+	if clock == nil {
+		panic(fmt.Sprintf("%s: clock must be non-nil", name))
 	}
 
 	bfr := &BoundedFrequencyRunner{
-		name:        name,
-		fn:          fn,
-		minInterval: minInterval,
-		maxInterval: maxInterval,
-		run:         make(chan struct{}, 1),
-		retry:       make(chan struct{}, 1),
-		timer:       timer,
-	}
-	if minInterval == 0 {
-		bfr.limiter = nullLimiter{}
-	} else {
-		// allow burst updates in short succession
-		qps := float32(time.Second) / float32(minInterval)
-		bfr.limiter = flowcontrol.NewTokenBucketRateLimiterWithClock(qps, burstRuns, timer)
+		name:          name,
+		fn:            fn,
+		minInterval:   minInterval,
+		retryInterval: retryInterval,
+		maxInterval:   maxInterval,
+		run:           make(chan struct{}, 1),
+		clock:         clock,
 	}
+	// The retry timer only fires when Retry() is called.
+	// Create the timer but do not start it.
+	bfr.retryIntervalTimer = bfr.clock.NewTimer(bfr.retryInterval)
+	bfr.retryIntervalTimer.Stop()
 	return bfr
 }
 
 // Loop handles the periodic timer and run requests.  This is expected to be
 // called as a goroutine.
 func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
-	klog.V(3).Infof("%s Loop running", bfr.name)
-	bfr.timer.Reset(bfr.maxInterval)
+	klog.V(3).Infof("%s Loop running", bfr.name) // nolint:logcheck
+	defer close(bfr.run)
+	bfr.minIntervalTimer = bfr.clock.NewTimer(bfr.minInterval)
+	defer bfr.minIntervalTimer.Stop()
+
+	bfr.maxIntervalTimer = bfr.clock.NewTimer(bfr.maxInterval)
+	defer bfr.maxIntervalTimer.Stop()
```

Comment:
couldn't you create the `retryTimer` here too to be more consistent?

Reply:
I did, but the problem is that it panics on the kube-proxy unit tests that does not start the loop but call retry ... so there was the option of running the loop on the kube-proxy unit test or initialize the stopped timer at creation time, I choose the later for simplicity to avoid having to start this runner on the unit tests

Reply:
fixed, we no longer use `Retry()`